src/rmp/src/Restructure.cpp

```diff
@@ -155,53 +160,138 @@ void Restructure::runABC()
-                        : static_cast<int>(Mode::DELAY_4));
-       opt_mode++) {
-    opt_mode_ = (Mode) opt_mode;
-    output_blif_file_name_ = work_dir_name_
-                             + std::string(block_->getConstName())
-                             + std::to_string(opt_mode) + "_crit_path_out.blif";
-    debugPrint(logger_,
-               RMP,
-               "remap",
-               1,
-               "Writing ABC script file {}",
-               abc_script_file);
-    if (writeAbcScript(abc_script_file)) {
-      std::string abc_command = std::string("yosys-abc < ") + abc_script_file;
-      if (logfile_ != "")
-        abc_command = abc_command + " > " + logfile_;
-      int ret = std::system(abc_command.c_str());
-      if (ret) {
+
+  debugPrint(logger_,
+             RMP,
+             "remap",
+             1,
+             "Running abc with number of threads = {}",
+             max_threads);
+
+  for (int curr_mode_idx = 0; curr_mode_idx < modes.size();) {
+    int max_parallel_runs = (max_threads < modes.size() - curr_mode_idx)
+                                ? max_threads
+                                : modes.size() - curr_mode_idx;
+
+    // Spawn ABC process(es)
+    for (int curr_thread = 0; curr_thread < max_parallel_runs; ++curr_thread) {
+      int temp_mode_idx = curr_mode_idx + curr_thread;
+      output_blif_file_name_
+          = work_dir_name_ + std::string(block_->getConstName())
+            + std::to_string(temp_mode_idx) + "_crit_path_out.blif";
+
+      opt_mode_ = modes[temp_mode_idx];
+
+      std::string abc_script_file = work_dir_name_
+                                    + std::to_string(temp_mode_idx)
+                                    + "ord_abc_script.tcl";
+
+      debugPrint(logger_,
+                 RMP,
+                 "remap",
+                 1,
+                 "Writing ABC script file {}",
+                 abc_script_file);
+      if (writeAbcScript(abc_script_file)) {
+        std::string abc_command = std::string("yosys-abc < ") + abc_script_file;
+        if (logfile_ != "")
+          abc_command
+              = abc_command + " > " + logfile_ + std::to_string(temp_mode_idx);
+
+        pid_t child_pid = fork();
+        if (child_pid == 0) { // Begin child
+          // Run in child process
+          int ret = execlp("sh", "sh", "-c", abc_command.c_str(), 0);
+          // Execution of command failed
+          logger_->warn(RMP,
+                        31,
+                        "Failed to run ABC with exit code {}. Please check the "
+                        "messages for details.",
+                        ret);
+          exit(ret);
+        } // End child
+
+        if (child_pid > 0) {
+          child_proc[temp_mode_idx] = child_pid;
+        } else if (child_pid < 0) {
+          logger_->warn(RMP,
+                        29,
+                        "Failed to create new ABC process. Please check the "
+                        "messages for details.");
```

Comment:
can we add why it failed? or at least the error code as the message above?

Reply:
This has to do with the OS, if the OS for some reason couldn't fork the parent process -1 will always be returned. I don't think there's any point in printing the -1 since it's constant during failure.

Reply:
"Failed to create new ABC process, could not fork parent process. Please check OS messages for details."?